File: day1/day1.cpp
#include <fstream>
#include <iostream>
#include <vector>
#include <algorithm>
#include <map>


void part1 (std::vector<int> vectora, std::vector<int> vectorb){
    std::sort(vectora.begin(), vectora.end());
    std::sort(vectorb.begin(), vectorb.end());
    int sum = 0;
    for (int i=0; i<vectora.size(); i++){
        sum += std::abs(vectora[i] - vectorb[i]);
    }
    std::cout << "Part 1:\n" << sum <<"\n";

}

void part2 (std::vector<int> vectora, std::vector<int> vectorb){
    std::map<int, int> map_vector2;

    for (int i=0; i<vectorb.size(); i++){
        
        if (map_vector2.find(vectorb[i]) != map_vector2.end()){
            map_vector2[vectorb[i]] += 1;

        }
        else{
            map_vector2[vectorb[i]] = 1;
        }
    }

    int sum = 0;
    for (int i=0; i<vectora.size(); i++){
        if (map_vector2.find(vectora[i]) != map_vector2.end()){
            sum += vectora[i] * map_vector2[vectora[i]];
        }
    }

    std::cout << "Part 2:\n" << sum <<"\n";


}

int main(){
    std::ifstream inputfile {"./day1_input.txt"};
    // std::ifstream inputfile {"./day1_input.txt"};

    std::string line;
    std::vector<int> vectora, vectorb;

    if (inputfile.is_open()){
        while (getline(inputfile, line)){   // rejects the last newline operator itself
            // std::cout << "reading line:" << line << "\n"; 

            std::size_t first_space = line.find_first_of(" ");
            std::size_t last_space = line.find_last_of(" ");
            
            int  a = std::stoi(line.substr(0, first_space));
            int  b = std::stoi(line.substr(last_space+1)); 
        
            vectora.push_back(a);
            vectorb.push_back(b);
        }
        inputfile.close();
    }
    
    part1(vectora, vectorb);
    std::cout << "-------------- \n";
    part2(vectora, vectorb);
    

    return 0;

}

File: day1/day1.py


def part1(list1, list2):
    list1.sort()
    list2.sort()
    
    abs_diff_list = []

    for i,_ in enumerate(list1):
        abs_diff_list.append(abs(list1[i] - list2[i]))

    print(sum(abs_diff_list))

def part2(list1, list2):

    dict_list2 = {}
    for item in list2:
        if item in dict_list2.keys():
            dict_list2[item] += 1
        else:
            dict_list2[item] = 1
    # print(dict_list2)

    sum_score = 0
    dict_keys = dict_list2.keys()
    for item in list1:
        if item in dict_keys:
            sum_score += item * dict_list2[item]

    print(sum_score)


def main():
    list1 = []
    list2 = []

    with open("./day1_input.txt") as f:
        lines = f.readlines()

    for line in lines:
        a, b = line.split('   ')
        b = b.rstrip('\n')
        list1.append(int(a))
        list2.append(int(b))
    
    print("Part 1:")
    part1(list1, list2)
    print("------------")
    print("Part 2:")
    part2(list1, list2)


if __name__ == "__main__":
    main()